File: C++/PalindromePartitioning.cpp
#include<vector>
using namespace std;

class Solution {
public:
	vector<vector<string>> all;
	bool isPalin(string s, int start, int end)
	{
		while(start < end)
		{
			if(s.at(start) != s.at(end))
				return false;
			start++;
			end--;
		}
		return true;
	}
	void findPalin(string s, int start, vector<string> v)
	{
		if(start == s.length())
		{
			all.push_back(v);
			return;
		}
		int len = s.length();
		for(int i = start; i < len; i++)
		{
			if(isPalin(s, start, i))
			{
				v.push_back(s.substr(start,i - start + 1));
				findPalin(s, i+1, v);
				v.pop_back();
			}
		}
	}
    vector<vector<string>> partition(string s) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		all.clear();
        vector<string> v;
		findPalin(s, 0, v);
		return all;
    }
};

File: JAVA/PalindromePartitioning.java
import java.util.ArrayList;

class Solution {
	ArrayList<ArrayList<String>> ret = new ArrayList<ArrayList<String>>();

	public ArrayList<ArrayList<String>> partition(String s) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ret.clear();
		if (s == null || s.length() == 0)
			return ret;
		ArrayList<String> list = new ArrayList<String>();
		solve(0, s.length(), s, list);
		return ret;
	}

	private void solve(int index, int len, String s, ArrayList<String> list) {
		if (index == len) {
			ret.add(new ArrayList<String>(list));
			return;
		}
		for (int i = index+1; i <= len; i++) {
			String sub = s.substring(index, i);
			if (isPalindrome(sub)) {
				list.add(sub);
				solve(i, len, s, list);
				list.remove(list.size() - 1);
			}
		}
	}

	private boolean isPalindrome(String s) {
		if (s == null)
			return false;
		int mid = s.length() / 2;
		for (int i = 0; i < mid; i++) {
			if (s.charAt(i) != s.charAt(s.length() - i -1))
				return false;
		}
		return true;
	}
}

public class PalindromePartitioning {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		String s = "abbab";
		Solution sol = new Solution();
		System.out.println(sol.partition(s));
	}

}
